dataset.py

```diff
@@ -137,3 +142,91 @@ def yearly_255_trees_partitioned(reference_csv_filename: str) -> PartitionedData
   print_split(partition_data)
   return partition_data
 
+#Utility function for randomly sampling a point around a sample site
+def is_valid_point(lat: float, lon: float, reference_isocape: raster.AmazonGeoTiff):
+  return True if raster.get_data_at_coords(reference_isocape, lon, lat, 0) else False
+
+# Pick a random point around (lat, lon) within max_distance_km. If edge_only is
+# true, only pick points exactly max_distance_km away from (lat, lon).
+def random_nearby_point(lat, lon, max_distance_km, edge_only=False):
+  # Pick a random angle pointing outward from the origin.
+  # 0 == North, 90 == East, 180 == South, 270 == West
+  angle = 360 * random.random()
+
+  # sqrt() is required for an equal radial distribution, otherwise samples
+  # cluster around origin.
+  dist = max_distance_km if edge_only else max_distance_km * math.sqrt(random.random())
+
+  # WGS-84 is the most accurate ellipsoidal model of Earth, but we should double
+  # check to make sure this matches the model used by our sample collectors.
+  point = distance.geodesic(
+      ellipsoid='WGS-84', kilometers=dist).destination((lat, lon), bearing=angle)
+  return point.latitude, point.longitude
+
+# Given a list of real_points, returns true if (lat, lon) is within threshold
+# of any of those points.
+def is_nearby_real_point(lat, lon, real_points, threshold_km):
+  for point, _ in real_points:
+    if distance.geodesic((lat, lon), point).km < threshold_km:
+      return True
+  return False
+
+
+#This function creates a dataset based on real samples adding a Fraud column
+def create_fraudulent_samples(real_samples_data, mean_iso,element,max_trusted_radius,max_fraud_radius,min_fraud_radius):
+  '''
+  This function creates a dataset based on real samples adding a Fraud column, where True represents a real lat/lon and False represents a fraudulent lat/lon
+  Input:
+  - real_samples_data: real samples
+  - element: element e.g d18O_cel
+  - mean_iso: isoscape averages
+  - max_trusted_radius, In km, the maximum distance from a real point where its value is still considered legitimate.
+  - max_fraud_radius: In km, the maximum distance from a real point to randomly sample a fraudalent coordinate.
+  - min_fraud_radius: In km, the minimum distance from a real point to randomly sample a fraudalent coordinate.
+  Output: 
+  - fake_data: pd.DataFrame with lat, long, isotope_value and fraudulent columns
+  '''
+
+  real_samples_data.dropna(subset=[element], how='all', inplace=True)
+  real_samples = real_samples_data.groupby(['lat','long'])[element]
+  real_samples_code = real_samples_data.groupby(['lat','long','Code'])[element]
+
+  count = 0
+  lab_samp = real_samples
+  #start number for fake code
+  fake_code_start = 5000
```

Comment:
Is this guaranteed to be unique i.e. can we say there will never be a mad5000 in our data, especially if we expect to get more data in the future?

I'd suggest changing Code formulation below to be unique to fraudulent samples i.e. f"fake_mad{fake_code_start+count}.

Reply:
'Code': f"fake_mad{count}"